File: dehaze-android/sdk/src/main/java/com/pei/dehaze/sdk/logger/LogEntry.java
package com.pei.dehaze.sdk.logger;

import com.google.gson.JsonObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * 单条前端日志条目。
 *
 * 字段规范见前端日志监控改造计划 §3.3（与 JS/Flutter 端对齐，service=client）。
 */
public class LogEntry {

    private final String timestamp;
    private final LogLevel level;
    private final String message;
    private final String app;
    private final String appVersion;
    private String url;
    private String userAgent;
    private String traceId;
    private Integer userId;
    private String errorType;
    private String errorSource;
    private String errorStack;
    private String method;
    private String path;
    private Integer status;
    private Double duration;
    private String code;

    public LogEntry(LogLevel level, String message, String app, String appVersion) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US);
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
        this.timestamp = sdf.format(new Date());
        this.level = level;
        this.message = message;
        this.app = app;
        this.appVersion = appVersion;
    }

    // ==================== setters ====================

    public LogEntry setUrl(String url) {
        this.url = url;
        return this;
    }

    public LogEntry setUserAgent(String userAgent) {
        this.userAgent = userAgent;
        return this;
    }

    public LogEntry setTraceId(String traceId) {
        this.traceId = traceId;
        return this;
    }

    public LogEntry setUserId(Integer userId) {
        this.userId = userId;
        return this;
    }

    public LogEntry setErrorType(String errorType) {
        this.errorType = errorType;
        return this;
    }

    public LogEntry setErrorSource(String errorSource) {
        this.errorSource = errorSource;
        return this;
    }

    public LogEntry setErrorStack(String errorStack) {
        this.errorStack = errorStack;
        return this;
    }

    public LogEntry setMethod(String method) {
        this.method = method;
        return this;
    }

    public LogEntry setPath(String path) {
        this.path = path;
        return this;
    }

    public LogEntry setStatus(Integer status) {
        this.status = status;
        return this;
    }

    public LogEntry setDuration(Double duration) {
        this.duration = duration;
        return this;
    }

    public LogEntry setCode(String code) {
        this.code = code;
        return this;
    }

    // ==================== getters ====================

    public String getTimestamp() { return timestamp; }
    public LogLevel getLevel() { return level; }
    public String getMessage() { return message; }
    public String getApp() { return app; }
    public String getAppVersion() { return appVersion; }
    public String getUrl() { return url; }
    public String getUserAgent() { return userAgent; }
    public String getTraceId() { return traceId; }
    public Integer getUserId() { return userId; }
    public String getErrorType() { return errorType; }
    public String getErrorSource() { return errorSource; }
    public String getErrorStack() { return errorStack; }
    public String getMethod() { return method; }
    public String getPath() { return path; }
    public Integer getStatus() { return status; }
    public Double getDuration() { return duration; }
    public String getCode() { return code; }

    /**
     * 序列化为 NDJSON 对象（仅输出非空字段）。
     */
    public JsonObject toJson() {
        JsonObject obj = new JsonObject();
        obj.addProperty("timestamp", timestamp);
        obj.addProperty("level", level.getLabel());
        obj.addProperty("message", message);
        obj.addProperty("app", app);
        obj.addProperty("app_version", appVersion);
        if (url != null && !url.isEmpty()) obj.addProperty("url", url);
        if (userAgent != null && !userAgent.isEmpty()) obj.addProperty("user_agent", userAgent);
        if (traceId != null && !traceId.isEmpty()) obj.addProperty("trace_id", traceId);
        if (userId != null) obj.addProperty("user_id", userId);
        if (errorType != null) obj.addProperty("error_type", errorType);
        if (errorSource != null) obj.addProperty("error_source", errorSource);
        if (errorStack != null) obj.addProperty("error_stack", errorStack);
        if (method != null) obj.addProperty("method", method);
        if (path != null) obj.addProperty("path", path);
        if (status != null) obj.addProperty("status", status);
        if (duration != null) obj.addProperty("duration", duration);
        if (code != null) obj.addProperty("code", code);
        return obj;
    }
}

File: dehaze_flutter/lib/core/logger/log_entry.dart
import 'dart:convert';

import 'logger.dart';

/// 单条前端日志条目。
///
/// 字段规范见前端日志监控改造计划 §3.3（与 JS SDK 对齐，service=client）。
class LogEntry {
  const LogEntry({
    required this.timestamp,
    required this.level,
    required this.message,
    required this.app,
    required this.appVersion,
    this.url,
    this.userAgent,
    this.traceId,
    this.errorType,
    this.errorSource,
    this.errorStack,
    this.method,
    this.path,
    this.status,
    this.duration,
    this.code,
    this.dedupCount,
  });

  /// ISO8601 UTC 时间戳
  final String timestamp;

  /// 日志级别
  final LogLevel level;

  /// 人读描述（≤2000 字符）
  final String message;

  /// 前端项目标识，固定 'flutter'
  final String app;

  /// 应用版本号
  final String appVersion;

  /// 当前路由/页面
  final String? url;

  /// 设备 User-Agent
  final String? userAgent;

  /// 与后端日志串联的 trace_id
  final String? traceId;

  /// 错误类型：dart
  final String? errorType;

  /// 错误来源：FlutterError / Zone / api_interceptor
  final String? errorSource;

  /// 完整堆栈字符串（≤8000 字符）
  final String? errorStack;

  /// HTTP 方法（API 失败日志）
  final String? method;

  /// 请求路径（API 失败日志，不含 query）
  final String? path;

  /// HTTP 状态码
  final int? status;

  /// 请求耗时（毫秒）
  final double? duration;

  /// 业务错误码
  final String? code;

  /// ERROR 去重汇总标记：10s 窗口内相同 fingerprint 的总命中次数（仅汇总条目携带，>1）
  final int? dedupCount;

  /// 序列化为 API 上报/落盘的 JSON map（NDJSON）。
  ///
  /// 仅输出非空字段，避免冗余。
  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'timestamp': timestamp,
      'level': level.label,
      'message': message,
      'app': app,
      'app_version': appVersion,
    };
    if (url != null && url!.isNotEmpty) map['url'] = url;
    if (userAgent != null && userAgent!.isNotEmpty) map['user_agent'] = userAgent;
    if (traceId != null && traceId!.isNotEmpty) map['trace_id'] = traceId;
    if (errorType != null) map['error_type'] = errorType;
    if (errorSource != null) map['error_source'] = errorSource;
    if (errorStack != null) map['error_stack'] = errorStack;
    if (method != null) map['method'] = method;
    if (path != null) map['path'] = path;
    if (status != null) map['status'] = status;
    if (duration != null) map['duration'] = duration;
    if (code != null) map['code'] = code;
    if (dedupCount != null) map['dedup_count'] = dedupCount;
    return map;
  }

  /// 序列化为 NDJSON 单行。
  String toNdjson() => jsonEncode(toJson());
}
